My Store Analytics: Custom range redacted view information

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12162
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds an informational alert to explain why visitors and conversion information are redacted from custom range tab.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Start with a WordPress.com or Jetpack-connected site,
2. On the analytics area on My Store, go to the Custom Range tab. If you don't have a Custom Range tab, make one first using the calendar-plus icon,
3. When the Custom Range tab, confirm that the Visitors and Conversion area at the top (below revenue) is redacted,
4. Tap the redacted view for both Visitors and Conversion, and ensure the alert like on the screenshot below is shown.
5. Select a value in the graph, so that now Visitors and Conversion is showing some number. Tap them, ensure the alert is not shown.
6. Switch to a Jetpack CP site,
7. Tap the redacted view for both Visitors and Conversion, and ensure the alert is not shown.
8. Switch to a non-Jetpack site,
9. Tap the view for both Visitors and Conversion, and ensure the alert is not shown.

## screenshot & video
<!-- Include before and after images or gifs when appropriate. -->

<img src="https://github.com/woocommerce/woocommerce-ios/assets/266376/c714a314-5b8b-4f95-94e9-169aad88abe5" width="42%" /> 

https://github.com/woocommerce/woocommerce-ios/assets/266376/950ab306-bdbf-4d81-986a-cdb60946b297  


---

### Additional case: handling for less than 2 days custom range.

When custom date range is in less than 2 days, we want to add this behavior:
1. When custom range is first created (in deselected state), show visits data.
2. When a value is selected, then hide visits data.
3. When a value is deselected, again show the visits data.

To test this:
1. Create a custom date range that is only 1 day apart,
2. When created, ensure that Visitors and Conversion are showing value.
3. Select a value in the graph, ensure that Visitors and Conversion are now redacted (simple redaction state).
4. Deselect the value, ensure that Visitors and Conversion are now shown again.
5. Update custom range to be 2 days apart,
6. When created, ensure that Visitors and Conversion are redacted (redacted due to custom range state),
7. Select a value in the graph, ensure that Visitors and Conversion are now shown.
8. Deselect the value, ensure that Visitors and Conversion are now redacted again.

Check video below for example:


https://github.com/woocommerce/woocommerce-ios/assets/266376/f23ff7ba-d019-49f7-8263-f1cddbd1bdd8

hafizrahman committed:
Add information alert for custom range redacted stats.

hafizrahman committed:
Show info fancy alert when redacted view is tapped.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/189">17.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
No need to recreate fancyAlerts.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift:
```diff
@@ -354,6 +359,30 @@ private extension StoreStatsV4PeriodViewController {
         updateStatsDataToDefaultStyles()
 
+        // Taps
+        if timeRange.isCustomTimeRange {
```
We should also check if the site is JCP or doesn't have Jetpack connection and not show the alert for those cases.

hafizrahman replied:
Thanks, I updated the logic in https://github.com/woocommerce/woocommerce-ios/pull/12226/commits/dd8989ccd0177e943a3554acc147ef88d7657afa to handle these two things:

1. Make sure non-Jetpack and Jetpack CP sites do not show the info alert when tapped,
2. Make sure info alert is also not shown when in custom range wpcom/Jetpack sites, and individual graph value is selected (e.g: when the visitors/conversion view are showing values).

I updated the test instructions above to include these.

itsmeichigo reviewed (commented):
This works great for sites with Jetpack connection through the Jetpack plugin, but we should not display the alert for JCP and non-JP sites (because the reason for the unavailable stats is different)

hafizrahman committed:
Update logic to not make info available on non wpcom/jetpack sites, and make it tappable only when in redacted state.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift:
```diff
@@ -80,10 +82,26 @@ final class StoreStatsV4PeriodViewController: UIViewController {
+    // To check whether the tab is showing the visitors and conversion views as redacted for custom range.
+    // This redaction is only shown on Custom Range tab with WordPress.com or Jetpack connected sites,
+    // while Jetpack CP sites has its own redacted for Jetpack state, and non-Jetpack sites simply has them empty.
+    private var isRedactedForCustomRangeShown: Bool {
+        guard timeRange.isCustomTimeRange,
+              let site = stores.sessionManager.defaultSite,
+              site.isJetpackConnected,
+              site.isJetpackThePluginInstalled else {
+            return false
+        }
+        return true
+    }
```
Great work reusing this check! 👏

itsmeichigo replied:
This is nit-picking but I'd suggest renaming this variable to `unavailableVisitStatsDueToCustomRange` to make it clearer.

itsmeichigo replied:
Also: if want to follow Hicham's idea here [p1709800424149309/1709727324.356419-slack-C03L1NF1EA3], we'll have to check the range length to not redact the site visit stats for 1-day range.

itsmeichigo reviewed (approved):
This works great now!

hafizrahman committed:
Refactor name to something better

hafizrahman committed:
Merge branch 'trunk'

hafizrahman committed:
Update redaction state during custom range creation to account for less than 2 days range.

In this case, we want to show the visit stats because they'll be accurate.

hafizrahman committed:
handle less than two days custom range behavior when selecting/deselecting value.

In this case, it should be the reverse of the "normal" behavior: redact data when selected, show when deselected.

hafizrahman committed:
Redaction state for less than 2 days should be .hidden

hafizrahman commented:
@itsmeichigo I added the handling for "less than two days" case in the last 3 commits, and updated the original post with the "Additional case: handling for less than 2 days custom range." section.

Please have a look for another review, thanks!

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -521,21 +521,26 @@ private extension StoreStatsAndTopPerformersViewController {
+
+        var siteVisitStatsMode: SiteVisitStatsMode
+        if site.isJetpackConnected && site.isJetpackThePluginInstalled {
+            siteVisitStatsMode = StatsTimeRangeV4.differenceInDays(startDate: startDate, endDate: endDate) == .lessThan2
+                ? .default
+                : .redactedDueToCustomRange
```
Nit:

```suggestion
            let differenceInDay = StatsTimeRangeV4.differenceInDays(startDate: startDate, endDate: endDate)
            siteVisitStatsMode = differenceInDay == .lessThan2 ? .default : .redactedDueToCustomRange
```

itsmeichigo reviewed (approved):
LGTM :shipit:

hafizrahman committed:
Refactor for better readability

Co-authored-by: Huong Do <hi.huongdo@gmail.com>
